Excessive time and memory usage by scatter

jbbieniekTT opened the pull request:
### Ticket
#22926

### Problem description
`ttnn.experimental.scatter_` takes way too much time/memory for pre-/post-processing

### What's changed
- converting input tensors into row-major
- suited kernels
- improved test cases

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [x] New/Existing tests provide coverage for changes

jbbieniekTT committed:
22926: row-major scatter

it's almost working, pushing to output doesn't work

jbbieniekTT committed:
22926: split rows

split rows for processing

bbradelTT reviewed (commented):
If this flips around the fatals, what happens to tiled tensors then?

Is the performance going to be worse now going the other way?

What is the performance before and after for row major and tiled tensors?

bbradelTT commented on ttnn/cpp/ttnn/operations/experimental/scatter/scatter.cpp:
```diff
@@ -24,34 +26,52 @@ Tensor pre_scatter_transform_tensor(
-    const Tensor transposed_tensor = reduction_common::perform_transpose(input_tensor, is_dim_last_idx, dim, -1);
-    Tensor transformed_tensor = reduction_common::transform_to_4d_tensor(transposed_tensor, is_rank_le_4d);
+    Tensor processed_tensor = input_tensor;
+    // if layout is tile, convert to row-major first
```
Please change to:
if layout is tile, convert to row-major first to avoid potentially large tensor size increase

jbbieniekTT committed:
22926: address PR comments

address PR comments

jbbieniekTT committed:
22926: fix bugs and address PR comments

- fix preallocated tensor bug
- fix validation bug
- fix shape transpose bug
- address PR comments

jbbieniekTT commented on tests/ttnn/unit_tests/operations/test_scatter.py:
```diff
@@ -14,40 +14,64 @@ def select_torch_dtype(ttnn_dtype):
     if ttnn_dtype is ttnn.int32:
-        return (
-            torch.int64
-        )  # !!! there is a strict requirement for the index tensor in Torch to be int64, and there is no int64 in ttnn
```
Done

aczajkowskiTT commented:
@jbbieniekTT Please include performance data for the new scatter

aczajkowskiTT commented:
Much better performance than before

jbbieniekTT committed:
22926: remove commented out code

remove commented out code

jbbieniekTT commented on ttnn/cpp/ttnn/operations/experimental/scatter/scatter.cpp:
```diff
@@ -18,40 +20,193 @@ namespace ttnn::operations::experimental {
+    //     TT_FATAL(output_tensor.storage_type() == StorageType::DEVICE, "Output tensor must be allocated on a
+    //     device."); TT_FATAL(!output_tensor.is_sharded(), "Sharded tensors are not supported - output_tensor is
+    //     sharded.");
+    // }
```
Removed

rfurko-tt commented on ttnn/cpp/ttnn/operations/experimental/scatter/device/kernels/dataflow/reader_scatter.cpp:
```diff
@@ -10,142 +10,71 @@
-        // release input tile + push resulting tile to the output
-        cb_push_back(output_cb, ONE_TILE);
-        cb_pop_front(input_cb, ONE_TILE);
+        volatile number_type& source_value = source_l1_read_ptr[index_in_index_chunk];
```
are you sure you need `volatile` here?

rfurko-tt commented on ttnn/cpp/ttnn/operations/experimental/scatter/device/kernels/dataflow/writer_scatter.cpp:
```diff
@@ -8,32 +8,43 @@
 void kernel_main() {
     constexpr auto ctas{get_ctas()};
 
+    const uint32_t output_buffer_address = get_arg_val<uint32_t>(0);
```
👍

rfurko-tt commented on ttnn/cpp/ttnn/operations/experimental/scatter/device/scatter_program_factory.cpp:
```diff
@@ -49,52 +55,72 @@ ScatterProgramFactory::cached_program_t ScatterProgramFactory::create(
+
+    const uint64_t L1_max_available_memory_bytes = device->l1_size_per_core();
+    constexpr uint64_t L1_reserve_bytes = (1 << 8) * (1 << 10);
+    const uint32_t input_and_output_max_chunk_size = 76800;
```
just curious, what does this number mean? :)

jbbieniekTT replied:
as per calculations, we can fit 4 rows (input/index/source/output) each of 4-byte datums in each core in its L1 memoryl. this enables us to cover 152600-element tensors in 12 swaps (152600 > 151936) as per kernel's implementation logic (4*2 for index/source + 2*2 for input/output), making it pretty efficient. this is going to get improved first off by another PR enabling work on one row among multiple Tensix, and this is probably not the only speed-up coming up in the nearest days

rfurko-tt commented on ttnn/cpp/ttnn/operations/experimental/scatter/device/scatter_program_factory.cpp:
```diff
@@ -108,68 +134,90 @@ ScatterProgramFactory::cached_program_t ScatterProgramFactory::create(
+    auto source_buffer_address = tensor_args.src_tensor.buffer()->address();
+    auto output_buffer_address = output_tensor.buffer()->address();
+    for (const auto& core : cores) {
+        GetRuntimeArgs(program, reader_kernel_id, core)[0] = input_buffer_address;
```
It's better to 
`auto& reader_runtime_args = GetRuntimeArgs(program, reader_kernel_id, core);` 

AFAIR `GetRuntimeArgs` is heavier call than expected

rfurko-tt replied:
no worries at all

you can go even further and do not call this operation for every call
here is an example: https://github.com/tenstorrent/tt-metal/blob/main/tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/cross_entropy_bw_program_factory.cpp#L454

rfurko-tt commented on ttnn/cpp/ttnn/operations/experimental/scatter/scatter.cpp:
```diff
@@ -71,84 +199,53 @@ Tensor ScatterOperation::invoke(
+        original_input_tensor_lshape,
+        original_layout);
+    if (opt_output.has_value()) {
+        ttnn::copy(output, *opt_output);
```
do we really need a copy here? 
wouldn't `opt_output = output;` be sufficient?

jbbieniekTT replied:
this actually doesn't work :(

rfurko-tt replied:
@philei-tt @sminakov-tt do you guys have an idea why it might not work? Am I missing something here? 
thanks in advance

jbbieniekTT replied:
removed for the time being

philei-tt replied:
@rfurko-tt Probably because operator= makes a shallow copy, and as I understood from @jbbieniekTT  left and right tensors have different layouts

rfurko-tt replied:
yeah, but is not shallow copy enough? 
don't understand why layout difference plays a role in this

jbbieniekTT committed:
22926: correction

rfurko-tt reviewed (approved):
LGTM
a few nits, but overall great shape of the PR! 

one more question, it seems like there are a lot of work done for properly supporting program cache (runtime arguments, override_runtime_args method impl).  Did you test it? 

If this change is part of the PR, please update PR description. Thanks in advance

jbbieniekTT committed:
22926: bail for unsupported yet cases

make scatter_ bail for unsupported yet cases
that depend on correct behaviour of transpose and to_layout
